src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -147,60 +156,103 @@ TryToDelegateFunctionCall(DistributedPlanningContext *planContext)
 		 * In pg12's planning phase empty FROMs are represented with an RTE_RESULT.
 		 * When we arrive here, standard_planner has already been called which calls
 		 * replace_empty_jointree() which replaces empty fromlist with a list of
 		 * single RTE_RESULT RangleTableRef node.
 		 */
-		if (list_length(joinTree->fromlist) == 1)
+		RangeTblRef *reference = linitial(joinTree->fromlist);
+
+		if (IsA(reference, RangeTblRef))
 		{
-			RangeTblRef *reference = linitial(joinTree->fromlist);
+			RangeTblEntry *rtentry = rt_fetch(reference->rtindex,
+											  planContext->query->rtable);
 
-			if (IsA(reference, RangeTblRef))
-			{
-				RangeTblEntry *rtentry = rt_fetch(reference->rtindex,
-												  planContext->query->rtable);
-				if (rtentry->rtekind != RTE_RESULT)
-				{
-					/* e.g. SELECT f() FROM rel */
-					return NULL;
-				}
-			}
-			else
+			if (rtentry->rtekind == RTE_FUNCTION)
 			{
 				/*
-				 * e.g. IsA(reference, JoinExpr). This is explicit join expressions
-				 * like INNER JOIN, NATURAL JOIN, ...
+				 * Look for a function in the FROM clause.
 				 */
+				fromFuncExpr = FunctionInFromClause(joinTree->fromlist,
+													planContext->query);
+			}
+			else if (rtentry->rtekind != RTE_RESULT)
+			{
+				/* e.g. SELECT f() FROM rel */
+				ereport(DEBUG4, (errmsg("FromList item is not empty")));
 				return NULL;
 			}
 		}
 		else
 		{
-			/* e.g. SELECT ... FROM rel1, rel2. */
-			Assert(list_length(joinTree->fromlist) > 1);
+			/*
+			 * e.g. IsA(reference, JoinExpr). This is explicit join expressions
+			 * like INNER JOIN, NATURAL JOIN, ...
+			 */
 			return NULL;
 		}
 	}
 
+	FuncExpr *targetFuncExpr = NULL;
 	List *targetList = planContext->query->targetList;
-	if (list_length(planContext->query->targetList) != 1)
+	int targetListLen = list_length(targetList);
+
+	if (targetListLen > 1)
 	{
 		/* multiple target list items */
 		return NULL;
 	}
+	else if (targetListLen == 1)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) linitial(targetList);
+		if (IsA(targetEntry->expr, FuncExpr))
+		{
+			/* function from the SELECT clause e.g. SELECT fn() FROM  */
+			targetFuncExpr = (FuncExpr *) targetEntry->expr;
+		}
+	}
 
-	TargetEntry *targetEntry = (TargetEntry *) linitial(targetList);
-	if (!IsA(targetEntry->expr, FuncExpr))
+	/*
+	 * Look for one of:
+	 * SELECT fn(...);
+	 * SELECT ... FROM fn(...);
+	 */
+	FuncExpr *funcExpr = NULL;
+	if (targetFuncExpr != NULL)
+	{
+		if (fromFuncExpr != NULL)
+		{
+			/* query is of the form: SELECT fn() FROM fn() */
+			return NULL;
+		}
+
+		/* query is of the form: SELECT fn(); */
+		funcExpr = targetFuncExpr;
+	}
+	else if (fromFuncExpr != NULL)
+	{
+		/* query is of the form: SELECT ... FROM fn(); */
+		funcExpr = fromFuncExpr;
+	}
+	else
 	{
-		/* target list item is not a function call */
+		/* query does not have a function call in SELECT or FROM */
+		ereport(DEBUG4, (errmsg("Either no function found or present"
+								" in both the SELECT and FROM list.")));
```

Comment:
I think our policy so far has been not to throw debug messages until we know we are dealing with a distributed function, to not clutter completely unrelated scenarios with debug messages.

Reply:
Removed